src/ray/object_manager/connection_pool.h

```diff
@@ -0,0 +1,148 @@
+  /// \param failure_callback The callback invoked if this method fails.
+  /// \return Status of invoking this method.
+  ray::Status GetSender(ConnectionType type, ClientID client_id,
+                        SuccessCallback success_callback,
+                        FailureCallback failure_callback);
+
+  /// Releasex a sender connection, allowing it to be used by another operation.
+  ///
+  /// \param type The type of connection.
+  /// \param conn The actual connection.
+  /// \return Status of invoking this method.
+  ray::Status ReleaseSender(ConnectionType type, SenderConnection::pointer conn);
+
+  /// Remove a sender connection. This is invoked if the connection is no longer
+  /// usable.
+  /// \param type The type of connection.
+  /// \param conn The actual connection.
+  /// \return Status of invoking this method.
+  // TODO(hme): Implement with error handling.
+  ray::Status RemoveSender(ConnectionType type, SenderConnection::pointer conn);
+
+ private:
+  /// A container type that maps ClientID to a connection type.
+  using SenderMapType =
+      std::unordered_map<ray::ClientID, std::vector<SenderConnection::pointer>,
+                         ray::UniqueIDHasher>;
+  using ReceiverMapType =
+      std::unordered_map<ray::ClientID, std::vector<std::shared_ptr<ReceiverConnection>>,
+                         ray::UniqueIDHasher>;
+
+  /// Adds a receiver for ClientID to the given map.
+  void Add(ReceiverMapType &conn_map, const ClientID &client_id,
+           std::shared_ptr<ReceiverConnection> conn);
+
+  /// Adds a sender for ClientID to the given map.
+  void Add(SenderMapType &conn_map, const ClientID &client_id,
+           SenderConnection::pointer conn);
+
+  /// Removes the given receiver for ClientID from the given map.
+  void Remove(ReceiverMapType &conn_map, const ClientID &client_id,
+              std::shared_ptr<ReceiverConnection> conn);
+
+  /// Returns the count of sender connections to ClientID.
+  uint64_t Count(SenderMapType &conn_map, const ClientID &client_id);
+
+  /// Removes a sender connection to ClientID from the pool of available connections.
+  /// This method assumes conn_map has available connections to ClientID.
+  SenderConnection::pointer Borrow(SenderMapType &conn_map, const ClientID &client_id);
+
+  /// Returns a sender connection to ClientID to the pool of available connections.
+  void Return(SenderMapType &conn_map, const ClientID &client_id,
+              SenderConnection::pointer conn);
+
+  /// Asynchronously obtain a connection to client_id.
+  /// If a connection to client_id already exists, the callback is invoked immediately.
+  ray::Status CreateConnection1(ConnectionType type, const ClientID &client_id,
+                                SuccessCallback success_callback,
+                                FailureCallback failure_callback);
+
+  /// Asynchronously create a connection to client_id.
+  ray::Status CreateConnection2(ConnectionType type, RemoteConnectionInfo info,
```

Comment:
Why do we need two versions of `CreateConnection` here? Let's try to make it one version.

Reply:
I refactored this to make the roles of these methods clearer.